subsys/bluetooth/mesh/cdb.c

```diff
@@ -0,0 +1,341 @@
+
+	return 0;
+}
+
+int bt_mesh_cdb_clear(void)
+{
+	int i;
+
+	atomic_clear_bit(bt_mesh_cdb.flags, BT_MESH_CDB_VALID);
+
+	for (i = 0; i < ARRAY_SIZE(bt_mesh_cdb.nodes); ++i) {
+		if (bt_mesh_cdb.nodes[i].addr != BT_MESH_ADDR_UNASSIGNED) {
+			bt_mesh_cdb_node_del(&bt_mesh_cdb.nodes[i], true);
+		}
+	}
+
+	for (i = 0; i < ARRAY_SIZE(bt_mesh_cdb.subnets); ++i) {
+		if (bt_mesh_cdb.subnets[i].net_idx != BT_MESH_KEY_UNUSED) {
+			bt_mesh_cdb_subnet_del(&bt_mesh_cdb.subnets[i], true);
+		}
+	}
+
+	for (i = 0; i < ARRAY_SIZE(bt_mesh_cdb.app_keys); ++i) {
+		if (bt_mesh_cdb.app_keys[i].net_idx != BT_MESH_KEY_UNUSED) {
+			bt_mesh_cdb_app_key_del(&bt_mesh_cdb.app_keys[i], true);
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
+		bt_mesh_store_cdb_net();
+	}
+
+	return 0;
+}
+
+struct bt_mesh_cdb_subnet *bt_mesh_cdb_subnet_alloc(u16_t net_idx)
+{
+	struct bt_mesh_cdb_subnet *sub;
+	int i;
+
+	if (bt_mesh_cdb_subnet_get(net_idx) != NULL) {
+		return NULL;
+	}
+
+	for (i = 0; i < ARRAY_SIZE(bt_mesh_cdb.subnets); ++i) {
+		sub = &bt_mesh_cdb.subnets[i];
+
+		if (sub->net_idx != BT_MESH_KEY_UNUSED) {
+			continue;
+		}
+
+		sub->net_idx = net_idx;
+
+		return sub;
+	}
+
+	return NULL;
+}
+
+void bt_mesh_cdb_subnet_del(struct bt_mesh_cdb_subnet *sub, bool store)
+{
+	BT_DBG("NetIdx 0x%03x store %u", sub->net_idx, store);
+
+	if (IS_ENABLED(CONFIG_BT_SETTINGS) && store) {
+		bt_mesh_clear_cdb_subnet(sub);
+	}
+
+	sub->net_idx = BT_MESH_KEY_UNUSED;
+	memset(sub->keys, 0, sizeof(sub->keys));
+}
+
+struct bt_mesh_cdb_subnet *bt_mesh_cdb_subnet_get(u16_t net_idx)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(bt_mesh_cdb.subnets); ++i) {
+		if (bt_mesh_cdb.subnets[i].net_idx == net_idx) {
+			return &bt_mesh_cdb.subnets[i];
+		}
+	}
+
+	return NULL;
+}
+
+struct bt_mesh_cdb_node *bt_mesh_cdb_node_alloc(const u8_t uuid[16], u16_t addr,
+						u8_t num_elem, u16_t net_idx)
+{
+	int i;
+
+	BT_DBG("");
+
+	if (addr == BT_MESH_ADDR_UNASSIGNED) {
+		addr = find_lowest_free_addr(num_elem);
+		if (addr == BT_MESH_ADDR_UNASSIGNED) {
+			return NULL;
+		}
+	} else if (addr_is_free(addr, num_elem, NULL) < 0) {
+		BT_DBG("Address range 0x%04x-0x%04x is not free", addr,
+		       addr + num_elem - 1);
+		return NULL;
+	}
+
+	for (i = 0; i < ARRAY_SIZE(bt_mesh_cdb.nodes); i++) {
+		struct bt_mesh_cdb_node *node = &bt_mesh_cdb.nodes[i];
+
+		if (node->addr == BT_MESH_ADDR_UNASSIGNED) {
+			memcpy(node->uuid, uuid, 16);
+			node->addr = addr;
+			node->num_elem = num_elem;
+			node->net_idx = net_idx;
+			return node;
+		}
+	}
+
+	return NULL;
+}
+
+void bt_mesh_cdb_node_del(struct bt_mesh_cdb_node *node, bool store)
+{
+	BT_DBG("Node addr 0x%04x store %u", node->addr, store);
+
+	if (IS_ENABLED(CONFIG_BT_SETTINGS) && store) {
+		bt_mesh_clear_cdb_node(node);
+	}
+
+	node->addr = BT_MESH_ADDR_UNASSIGNED;
+	memset(node->dev_key, 0, sizeof(node->dev_key));
+
+	/* FIXME:
+	 * Is this something that CDB should care about?
+	 * How to handle it correctly?
+	 */
+	bt_mesh_rpl_reset();
```

Comment:
I have a feeling it shouldn't, i.e. it should be a higher level policy that does "configurer/provisioner as node in the network" actions rather than this being baked into the CDB implementation itself. Or do you have other places in the implementation where these two mix with each other?

Reply:
Yea I just noted that when I removed a node manually from the cdb, ie with `cdb-node-del` from the mesh-shell, and then re-used that address I got some replay warnings for that new node. Maybe the rpl should be cleared for an address when that node is provisioned?

Reply:
This is actually somewhat described in `3.10.7 Node Removal procedure` in the Mesh Profile specification.

> After a node is removed from a network, its unicast addresses may be reused by a Provisioner. A Provisioner shall only reuse these addresses after the current IV Index (at the time of removal) has been updated (see Section 3.10.5) in order to enable the SEQ numbers to be reused.

This is currently not followed in the implementation but I'll remove the `bt_mesh_rpl_reset` call as there is a correct way to handle it.